Applications/Spire/Source/Ui/TextAreaBox.cpp

```diff
@@ -0,0 +1,627 @@
+        if(!last_line.endsWith(ELLIPSES_CHAR)) {
+          if(last_line.endsWith('\n')) {
+            last_line.remove(last_line.size() - 1, 1);
+          }
+          last_line.append(ELLIPSES_CHAR);
+        }
+        m_text_edit->blockSignals(true);
+        m_text_edit->setText(lines.join(""));
+        m_text_edit->blockSignals(false);
+      }
+    }
+  } else if(m_text_edit->toPlainText() != m_model->get_current()) {
+    m_text_edit->setText(m_model->get_current());
+  }
+  update_text_alignment();
+  update_document_line_height();
+}
+
+void TextAreaBox::update_document_line_height() {
+  apply_block_formatting([=] (const auto& block) {
+    auto cursor = m_text_edit->textCursor();
+    cursor.setPosition(block.position());
+    m_text_edit->setTextCursor(cursor);
+    auto block_format = cursor.blockFormat();
+    block_format.setLineHeight(m_computed_line_height,
+      QTextBlockFormat::FixedHeight);
+    cursor.setBlockFormat(block_format);
+    m_text_edit->setTextCursor(cursor);
+  });
+}
+
+void TextAreaBox::update_layout() {
+  update_text_edit_width();
+  m_stacked_widget->setMinimumSize(size() - get_border_size() -
+    get_padding_size());
+  m_stacked_widget->adjustSize();
+  update_display_text();
+  update_placeholder_text();
+}
+
+void TextAreaBox::update_line_height() {
+  m_computed_line_height = static_cast<int>(m_text_edit->font().pixelSize() *
+    *m_text_edit_styles.m_line_height);
+  m_placeholder->set_line_height(m_computed_line_height);
+  m_scroll_box->get_vertical_scroll_bar().set_line_size(
+    m_computed_line_height);
+  update_display_text();
+}
+
+void TextAreaBox::update_placeholder_text() {
+  if(is_placeholder_shown()) {
+    m_placeholder->set_text(m_placeholder_text);
+    m_placeholder->setFixedSize(size() - get_border_size() -
+      get_padding_size());
+    m_stacked_widget->adjustSize();
+    m_placeholder->show();
+  } else {
+    m_placeholder->hide();
+  }
+}
+
+void TextAreaBox::update_text_alignment() {
+  apply_block_formatting(
+    [=, alignment = m_text_edit_styles.m_alignment] (const auto& block) {
+      auto cursor = m_text_edit->textCursor();
+      cursor.setPosition(block.position());
+      m_text_edit->setTextCursor(cursor);
+      m_text_edit->setAlignment(m_text_edit_styles.m_alignment);
+    });
+}
+
+void TextAreaBox::update_text_edit_width() {
+  auto border_size = get_border_size();
+  auto padding_size = get_padding_size();
+  if(!is_read_only() &&
+      m_text_edit->document()->size().toSize().height() >
+      height() - padding_size.height() - border_size.height()) {
+    m_text_edit->setFixedWidth(width() -
+      m_scroll_box->get_vertical_scroll_bar().sizeHint().width() -
+      border_size.width() - padding_size.width());
+  } else {
+    m_text_edit->setFixedWidth(width() - border_size.width() -
+      padding_size.width());
+  }
+}
+
+void TextAreaBox::on_current(const QString& current) {
+  if(m_text_edit->toPlainText() != current) {
```

Comment:
What does this check do? It's not cheap comparing two strings, by itself it's an ```O(N)``` and when you consider that typing happens character by character this makes typing an ```O(N^2)``` operation.